katdal/test/test_datasources.py

```diff
@@ -0,0 +1,113 @@
+            'weights_channel': ramp(shape[:-1], dtype=np.float32)}
+    ddata = {k: to_dask_array(array) for k, array in data.items()}
+    chunk_info = {k: {'chunks': darray.chunks, 'dtype': darray.dtype,
+                      'shape': darray.shape} for k, darray in ddata.items()}
+    push = [store.put_dask_array(store.join(base_name, k), darray)
+            for k, darray in ddata.items()]
+    da.compute(push)
+    return data, chunk_info
+
+
+class TestChunkStoreVisFlagsWeights(object):
+    """Test the :class:`ChunkStoreVisFlagsWeights` dataset store."""
+
+    @classmethod
+    def setup_class(cls):
+        cls.tempdir = tempfile.mkdtemp()
+
+    @classmethod
+    def teardown_class(cls):
+        shutil.rmtree(cls.tempdir)
+
+    def test_construction(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb1'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        vfw = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+        weights = data['weights'] * data['weights_channel'][..., np.newaxis]
+        # Check that data is as expected when accessed via VisFlagsWeights
+        assert_equal(vfw.shape, data['correlator_data'].shape)
+        assert_array_equal(vfw.vis.compute(), data['correlator_data'])
+        assert_array_equal(vfw.flags.compute(), data['flags'])
+        assert_array_equal(vfw.weights.compute(), weights)
+
+    def test_missing_chunks(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb2'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        # Delete a random chunk in each array of the dataset
+        missing_chunks = {}
+        for array, info in chunk_info.items():
+            array_name = store.join(base_name, array)
+            slices = da.core.slices_from_chunks(info['chunks'])
+            index = np.random.randint(len(slices))
+            missing_chunks[array] = slices[index]
+            chunk_name, shape = store.chunk_metadata(array_name, slices[index])
+            os.remove(os.path.join(store.path, chunk_name) + '.npy')
+        vfw = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+        # Check that missing chunks have been replaced by zeroes
+        assert_array_equal(vfw.vis[missing_chunks['correlator_data']], 0.)
+        assert_array_equal(vfw.weights[missing_chunks['weights']], 0.)
+        assert_array_equal(vfw.weights[missing_chunks['weights_channel']], 0.)
+        # Check that (only) missing chunks have been flagged as 'data lost'
+        expected = np.zeros_like(vfw.flags)
+        expected[missing_chunks['correlator_data']] |= 8
+        expected[missing_chunks['weights']] |= 8
+        expected[missing_chunks['weights_channel']] |= 8
+        assert_array_equal(vfw.flags & 8, expected)
```

Comment:
Should check all the bits, not just the data_lost bit, to ensure that existing flags are preserved where no data is lost.

Reply:
Done (same for vis and weights).